### PR TITLE
Add PKCE support

### PR DESCRIPTION
**Pull request by Kehrlann**
This is a first stab at PKCE implementation - see #45 . The idea is to validate the structure of how to achieve PKCE validation, and then start iterating from there.

Notes:
- The SHA256 code verification was stolen directly from spring-security, see `org.springframework.security.oauth2.client.web.DefaultOAuth2AuthorizationRequestResolver#createHash` . We might want to make that public in spring-security at some point ?
- We consider a client is _public client_ when the principal is null but it has client ID.

### Authorization endpoint

- [x] RFC 7636 - 4.4 Server Returns the Code ; validate code_challenge & code_challenge_method
- [x] CODE_CHALLENGE is required ; CODE_CHALLENGE_METHOD can be null (will default to plain)
- [x] Only trigger PKCE checks for public clients (`registeredClient.clientSecret` is blank)
- [ ] Consider what is a "public" client - maybe we could mark some "redirectUris" to require PKCE vali
- [x] Add tests for `OAuth2AuthorizationEndpointFilter`

### Token endpoint

- [x] Validate `code_verifier` for both "plain" and "S256" methods (for public clients)
- [x] If `code_challenge_method` is null, default to plain
- [x] Polish changes to `OAuth2AuthorizationCodeAuthenticationProvider` (null checks, try-catches)
- [x] In `OAuth2AuthorizationCodeAuthenticationProvider.authenticate`, authenticate public client - See RFC 6749 - 3.2.1.  Token Endpoint > Client Authentication [1]

### Integration tests

- [x] Add integration test for PKCE

[1] In the "authorization_code" "grant_type" request to the token endpoint, an unauthenticated client MUST send its "client_id" to prevent itself from inadvertently accepting a code intended for a client with a different "client_id". This protects the client from substitution of the authentication code.  (It provides no additional security for the protected resource.)

### CODE REVIEW

**Comment by Kehrlann** [comment]
I started adding tests for `OAuth2AuthorizationCodeAuthenticationProvider`. Two of them are failing, they show the need for `OAuth2AccessTokenAuthenticationToken` to support public clients - today, only private clients are supported.

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/PkceCodeChallengeMethodNames.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.server.authorization;
+
+// https://tools.ietf.org/html/rfc7636#section-6.2.2
+// PKCE Code Challenge Method Registry > Initial Registry Contents
+// Method names are case-sensistive
+public interface PkceCodeChallengeMethodNames {
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/PkceCodeChallengeMethodNames.java:21]: Remove and add these as `private static` constants in `OAuth2AuthorizationCodeAuthenticationProvider`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationToken.java [hunk 2]
```diff
@@ -48,30 +51,39 @@
 	 * @param redirectUri the redirect uri
 	 */
 	public OAuth2AuthorizationCodeAuthenticationToken(String code,
-			Authentication clientPrincipal, @Nullable String redirectUri) {
+			Authentication clientPrincipal, @Nullable Map<String, Object> additionalParameters,
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationToken.java:54]: Please move `additionalParameters` as last argument for both constructors.

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AccessTokenAuthenticationToken.java [hunk 3]
```diff
@@ -60,9 +61,23 @@ public OAuth2AccessTokenAuthenticationToken(RegisteredClient registeredClient,
 		this.accessToken = accessToken;
 	}
 
+	/**
+	 * Constructs an {@code OAuth2AccessTokenAuthenticationToken} for public clients, using the provided parameters.
+	 *
+	 * @param clientId the client id
+	 * @param accessToken the access token
+	 */
+	public OAuth2AccessTokenAuthenticationToken(String clientId, OAuth2AccessToken accessToken) {
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AccessTokenAuthenticationToken.java:70]: Please remove this constructor as the existing constructor may be used. For the `Authentication clientPrincipal` argument, use `new OAuth2ClientAuthenticationToken(RegisteredClient)`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java [hunk 4]
```diff
@@ -209,9 +213,27 @@ public Authentication convert(HttpServletRequest request) {
 				throwError(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2ParameterNames.REDIRECT_URI);
 			}
 
+			// code_verifier (REQUIRED for public clients)
+			if (clientPrincipal == null) {
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java:217]: Let's move this logic under `if (StringUtils.hasText(clientId)) {`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/PkceCodeChallengeMethodNames.java (hunk 1 above)
Comment (Kehrlann) [reply core/src/main/java/org/springframework/security/oauth2/server/authorization/PkceCodeChallengeMethodNames.java:21]: I don't know what's the best place to put those. Not sure this should go in OAuth2AuthorizationCodeAuthenticationProvider: they are used in multiple places.
- The CODE_VERIFIER constant is used by `OAuth2TokenEndpointFilter.AuthorizationCodeAuthenticationConverter` ; as it is a required parameter for public clients.
- The CODE_CHALLENGE and CODE_CHALLENGE_METHOD should also be checked in the Authorization endpoint, as per [RFC 7636](https://tools.ietf.org/html/rfc7636#section-4.4.1). For public clients, these parameters are required in the authorization request. Also, checking whether the server supports the transformation (for S256) should be done in the authorization endpoint. Relevant paragraph from the spec:

```
4.  Protocol
4.4.  Server Returns the Code
4.4.1.  Error Response
If the server requires Proof Key for Code Exchange (PKCE) by OAuth
public clients and the client does not send the "code_challenge" in
the request, the authorization endpoint MUST return the authorization
error response with the "error" value set to "invalid_request".  The
"error_description" or the response of "error_uri" SHOULD explain the
nature of error, e.g., code challenge required.
If the server supporting PKCE does not support the requested
transformation, the authorization endpoint MUST return the
authorization error response with "error" value set to
"invalid_request".  The "error_description" or the response of
"error_uri" SHOULD explain the nature of error, e.g., transform
algorithm not supported.
```

**Comment by Kehrlann** [comment]
The PR is ready for review. Still to be decided: how do we identify public clients in `OAuth2AuthorizationEndpointFilter` ?
1. When the `registeredClient` has no `clientSecret`, its' a 100% public client, that's currently covered.
2. However, the registered client could have a secret and still be used as a public client. Something to explore could be: some redirect uris of a client could be tagged with "require PKCE".

**Comment by uhhhh2** [comment]
Thoughts on letting all clients (public and confidential) that use authorization code also use PKCE (confidential will have both secret & PKCE), but only require PKCE for public clients. (those without client secrets) like in https://github.com/spring-projects/spring-security/issues/6548 ?

**Comment by Kehrlann** [comment]
With the current proposed implementation, confidential clients may send PKCE parameters. However, those parameters will not be validated. Not sure if we should put this in v1 ?

It makes sense to validate whatever PKCE params confidential clients send, especially if we're allowing confidential clients to do it in spring-security.

**Comment by Kehrlann** [comment]
Rebased on top of master. For info, I have only checked the build & ran the tests on conflicting commits.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AccessTokenAuthenticationToken.java [hunk 5]
```diff
@@ -38,9 +38,10 @@
  */
 public class OAuth2AccessTokenAuthenticationToken extends AbstractAuthenticationToken {
 	private static final long serialVersionUID = SpringSecurityCoreVersion2.SERIAL_VERSION_UID;
-	private final RegisteredClient registeredClient;
-	private final Authentication clientPrincipal;
 	private final OAuth2AccessToken accessToken;
+	private RegisteredClient registeredClient;
+	private Authentication clientPrincipal;
+	private String clientId;
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AccessTokenAuthenticationToken.java:44]: `clientId` is not used. I think this was left over from previous updates? Please revert the changes in this file.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java [hunk 6]
```diff
@@ -58,6 +64,8 @@
  * @see <a target="_blank" href="https://tools.ietf.org/html/rfc6749#section-4.1.3">Section 4.1.3 Access Token Request</a>
  */
 public class OAuth2AuthorizationCodeAuthenticationProvider implements AuthenticationProvider {
+	private static final String PKCE_METHOD_PLAIN = "plain";
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java:67]: Please remove these 2 constants and use inline with code given that it's used in one spot only.

**Review by jgrandja — CHANGES_REQUESTED** [review]
Thanks for the updates @Kehrlann! I left a few comments inline.
Also, can you please squash the commits on next update.
Thanks.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2AuthorizationEndpointFilter.java [hunk 7]
```diff
@@ -72,6 +69,10 @@
 	 * The default endpoint {@code URI} for authorization requests.
 	 */
 	public static final String DEFAULT_AUTHORIZATION_ENDPOINT_URI = "/oauth2/authorize";
+	private static final String OUATH2_SPEC_ERROR_URL = "https://tools.ietf.org/html/rfc6749#section-4.1.2.1";
```
[the remark this answers is not in the document]
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2AuthorizationEndpointFilter.java:72]: Done. The `PKCE_SPEC_ERROR_URL` is used three times, in a "createError" method. The only difference is the PkceParameterName.

```java
createError(
  OAuth2ErrorCodes.INVALID_REQUEST, 
  PkceParameterNames.CODE_CHALLENGE_METHOD,
  "https://tools.ietf.org/html/rfc7636#section-4.4.1");
```

I'm tempted to either keep `PKCE_SPEC_ERROR_URL` or to re-introduce a `createPkceError(String parameterName)`, ... but we explicitly removed that method in a previous iteration.

Thoughts ?

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationToken.java [hunk 8]
```diff
@@ -39,39 +40,43 @@
 	private Authentication clientPrincipal;
 	private String clientId;
 	private String redirectUri;
+	private final Map<String, Object> additionalParameters;
 
 	/**
 	 * Constructs an {@code OAuth2AuthorizationCodeAuthenticationToken} using the provided parameters.
-	 *
```
[the remark this answers is not in the document]
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationToken.java:45]: Woops, auto-formatting mishap. Is there an IDEA config somewhere I could use so this doesn't happen again ?

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java [hunk 9]
```diff
@@ -198,14 +201,22 @@ public Authentication convert(HttpServletRequest request) {
...
+				clientId = parameters.getFirst(OAuth2ParameterNames.CLIENT_ID);
+				if (!StringUtils.hasText(clientId) ||
+						parameters.get(OAuth2ParameterNames.CLIENT_ID).size() != 1) {
 					throwError(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2ParameterNames.CLIENT_ID);
 				}
-			} else {
-				clientPrincipal = SecurityContextHolder.getContext().getAuthentication();
+
+				// code_verifier (REQUIRED for public clients)
+				String codeVerifier = parameters.getFirst(PkceParameterNames.CODE_VERIFIER);
```
[the remark this answers is not in the document]
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java:215]: I'm wondering if we should not leave it here. The `CODE_VERIFIER` is a required parameter for public clients using the token endpoint. Failure to provide it yields an `INVALID_REQUEST` [RFC 7636 - 4.5](https://tools.ietf.org/html/rfc7636#section-4.5)

First, for consistency that'd be nice: 
- "required" params are checked in the `OAuth2EndpointFilters`, with `INVALID_REQUEST` as error when a param is missing.
- the `OAuth2AuthorizationCodeAuthenticationProvider` seems to be throwing only `INVALID_CLIENT` or `INVALID_GRANT` errors.

Second, I somewhat like having both the clientId and codeVerifier checked in the same block, showing what the required params are for un-authenticated clients.

Thoughts ?

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2AuthorizationEndpointFilter.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jgrandja) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2AuthorizationEndpointFilter.java]: Ok let's keep the constant and rename it to `PKCE_ERROR_URI`

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (jgrandja) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java:215]: Good points. Ok I agree we should check for required parameter(s) in the incoming request and throw `INVALID_REQUEST` when missing. This is consistent in all `Filter`'s. Let's keep it.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationToken.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (jgrandja) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationToken.java:45]: I don't have anything special configured in my IDEA. As a general practise, I always review the changes before I commit so that I can confirm no other changes were made inadvertently.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java [hunk 10]
```diff
@@ -85,29 +93,30 @@ public Authentication authenticate(Authentication authentication) throws Authent
 				(OAuth2AuthorizationCodeAuthenticationToken) authentication;
 
 		OAuth2ClientAuthenticationToken clientPrincipal = null;
+		RegisteredClient registeredClient = null;
 		if (OAuth2ClientAuthenticationToken.class.isAssignableFrom(authorizationCodeAuthentication.getPrincipal().getClass())) {
 			clientPrincipal = (OAuth2ClientAuthenticationToken) authorizationCodeAuthentication.getPrincipal();
-		}
-		if (clientPrincipal == null || !clientPrincipal.isAuthenticated()) {
+			registeredClient = clientPrincipal.getRegisteredClient();
+		} else if (authorizationCodeAuthentication.getPrincipal() instanceof String) {
```
[the remark this answers is not in the document]
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java:98]: So far, the principal for the `OAuth2AuthorizationCodeAuthenticationToken` is either `OAuth2ClientAuthenticationToken` or null (and in that case, there's a client ID). So for now we can do:

```java
/**
 * Returns the client id
 *
 * @return the client id
 */
public String getClientId() {
	if (this.clientPrincipal != null && OAuth2ClientAuthenticationToken.class.isAssignableFrom(this.clientPrincipal.getClass())) {
		return (String) this.clientPrincipal.getPrincipal();
	} else {
		return this.clientId;
	}
}
```

But I'm a bit worried about what happens if the principal is something else than an `OAuth2ClientAuthenticationToken` ; for now we just return null.

Thoughts ?

[the remark this answers is not in the document]
Comment (jgrandja) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java:98]: I think it makes sense to provide a getter for `clientId` given we already provide one for `code` and `redirectUri`. I think we should set the `clientId` in the constructor that accepts `Authentication clientPrincipal` to simplify `getClientId()` and the signature should be `public @Nullable String getClientId()`.

FYI, `OAuth2ClientAuthenticationToken` will change as soon as we add support for other client authentication methods, which will ultimately change `OAuth2AuthorizationCodeAuthenticationToken` and likely other parts in the code. I also feel that `OAuth2AuthorizationCodeAuthenticationToken` should probably accept `OAuth2ClientAuthenticationToken` instead of `Authentication` but let's leave it as-is for now.

The changes suggested will be fine for this iteration but like I said things will change soon enough.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java [hunk 11]
```diff
@@ -116,6 +125,41 @@ public Authentication authenticate(Authentication authentication) throws Authent
...
+		if (clientPrincipal == null) {
+			// RFC 7636 (PKCE)
+			// 4.6.  Server Verifies code_verifier before Returning the Tokens
+
+			String codeChallenge;
+			Object codeChallengeParameter = authorizationRequest
+					.getAdditionalParameters()
+					.get(PkceParameterNames.CODE_CHALLENGE);
+			if (codeChallengeParameter == null) {
+				throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes.INVALID_GRANT));
```
[the remark this answers is not in the document]
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java]: One last question here: if the client requires PKCE (`registeredClient.requiredProofKey`), that means the auth endpoint **must** require a `code_challenge`. However, we have no hard guarantee, in the Token endpoint, that his was actually the case.  

Should we:
1. Assume that the Authorization endpoint will never let a request through without `code_challenge` when `requireProofKey` is true, and ignore that case here.
1. Be paranoid and guard against the case where the Authorization endpoint didn't properly check the REQUIRED `code_challenge`, and did let a request through without a `code_challenge` ?

**Comment by Kehrlann** [comment]
Included all corrections, squashed. There's still one open question, and we should be almost good to go.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (jgrandja) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java]: Both points are valid actually.

Point 1 - the Authorization endpoint should error when `code_challenge` is missing and `requireProofKey` is `true`

Point 2 - there should be a test that checks for `code_challenge` when `requireProofKey` is `true`. However, I also think as an extra precautionary check, we could also (double) check for `code_challenge` when `requireProofKey` is `true`.

[the remark this answers is not in the document]
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java]: Perfect, thanks 👌

Point 1 is already addressed, see `OAuth2AuthorizationEndpointFilter`, lines 202-205.

```java
		} else if (registeredClient.getClientSettings().requireProofKey()) {
			OAuth2Error error = createError(OAuth2ErrorCodes.INVALID_REQUEST, PkceParameterNames.CODE_CHALLENGE, PKCE_ERROR_URI);
			sendErrorResponse(request, response, error, stateParameter, redirectUri);
			return;
		}
```

I just pushed the correction for point 2, in `OAuth2AuthorizationCodeAuthenticationProvider`, lines 150-152.

```java
		} else if (registeredClient.getClientSettings().requireProofKey()){
			throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes.INVALID_GRANT));
		}
```

**Comment by jgrandja** [comment]
Thanks for the great work @Kehrlann ! This is now in master!

FYI, I added a polish commit with some minor updates.

Also, I discovered an issue when reviewing the integration test `OAuth2PkceTests`. The test revealed an issue with the implementation so it needed to be re-worked. Ultimately, the client needs to be authenticated before it hits the token endpoint. Therefore, I needed to move the PKCE authentication logic from `OAuth2AuthorizationCodeAuthenticationProvider` to `OAuth2ClientAuthenticationProvider`.

Apologies, as I missed this during the review process. I decided to go ahead an apply the change. Take a look at the commit  5c31fb1 and let me know if you have any questions.

Thanks again for all the great work!
